[JSC] `Intl.DurationFormat#resolvedOptions().fractionalDigits` should be `undefined` if `fractionalDigits` is `undefined`.

sosukesuzuki opened the pull request:
#### a43717459fdb2af1b6dee3dbdb68488943c17710
<pre>
[JSC] `Intl.DurationFormat#resolvedOptions().fractionalDigits` should be `undefined` if `fractionalDigits` is `undefined`.
<a href="https://bugs.webkit.org/show_bug.cgi?id=274733">https://bugs.webkit.org/show_bug.cgi?id=274733</a>

Reviewed by Ross Kirsling.

According to the spec[1], when creating a `DurationFormat` instance with
`new Intl.DurationFormat(locales, { fractionalDigits: undefined });`,
`Intl.DurationFormat#resolvedOptions().fractionalDigits` should be `undefined`. However, in the
current JSC, it is `0`. This was occurring because, according to the spec[2], the
`fallback` argument of the `GetNumberOption` abstract operation should be `undefined`, but we were
passing `0`.

This patch makes the following changes:

- Pass `numeric_limits&lt;unsigned&gt;::max()`, representing `undefined`, to the `fallback` argument of
the `intlNumberOption`
function, which corresponds to the `GetNumberOption` abstract operation, when initializing
`DurationFormat` instances.
- In the `DurationFormat#resolvedOption` function, set the `fractionalDigits` field to
`jsUndefined()` if `m_fractionalDigits` is `numeric_limits&lt;unsigned&gt;::max()`.

[1]: <a href="https://tc39.es/proposal-intl-duration-format/#sec-Intl.DurationFormat.prototype.resolvedOptions">https://tc39.es/proposal-intl-duration-format/#sec-Intl.DurationFormat.prototype.resolvedOptions</a>
[2]: <a href="https://tc39.es/proposal-intl-duration-format/#sec-Intl.DurationFormat">https://tc39.es/proposal-intl-duration-format/#sec-Intl.DurationFormat</a>

* JSTests/stress/intl-durationformat-fractionaldigits-undefined.js: Added.
(shouldBe):
* JSTests/stress/intl-durationformat.js:
(test):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/IntlDurationFormat.cpp:
(JSC::IntlDurationFormat::initializeDurationFormat):
(JSC::collectElements):
(JSC::IntlDurationFormat::resolvedOptions const):

Canonical link: <a href="https://commits.webkit.org/279503@main">https://commits.webkit.org/279503@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc7e5b5a24e0bf65928122b7b14b465db43e214a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53668 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33033 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6184 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56949 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4394 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40517 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4227 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/43478 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2868 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55765 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/31251 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46397 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24612 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/28078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3717 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2549 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/47028 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49800 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3895 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58543 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/53180 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28831 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/3940 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/30030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46562 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/50229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11694 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/30963 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/65484 "Built successfully") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/29807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12476 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on JSTests/test262/expectations.yaml:
There are a couple of new failures. This is because `resolvedOptions().fractionalDigits` is now `undefined`, which changes the harness behavior of test262 (ref: https://github.com/tc39/test262/blob/e394ebb6fce16bbba916694b6bdc5d9efbf88410/harness/testIntl.js#L2629 ). The behavior of `Intl.DurationFromat#format` has not changed.

rkirsling commented on JSTests/test262/expectations.yaml:
```diff
@@ -1072,17 +1069,23 @@ test/intl402/DurationFormat/prototype/format/numeric-hour-with-zero-minutes-and-
-  strict mode: 'Test262Error: Duration is {"seconds":9007199254740991,"milliseconds":9007199254740991,"microseconds":9007199254740991,"nanoseconds":9007199254740991} Expected SameValue(«0:00:9,016,215,470,202,186», «0:00:9,016,215,470,202,185») to be true'
+  default: 'Test262Error: Duration is {"seconds":10000000,"nanoseconds":1} Expected SameValue(«0:00:10,000,000», «0:00:10,000,000.000000001») to be true'
+  strict mode: 'Test262Error: Duration is {"seconds":10000000,"nanoseconds":1} Expected SameValue(«0:00:10,000,000», «0:00:10,000,000.000000001») to be true'
+test/intl402/DurationFormat/prototype/format/style-digital-en.js:
+  default: 'Test262Error: Assert DurationFormat format output using digital style option Expected SameValue(«1 yr, 2 mths, 3 wks, 3 days, 4:05:06», «1 yr, 2 mths, 3 wks, 3 days, 4:05:06.007008009») to be true'
+  strict mode: 'Test262Error: Assert DurationFormat format output using digital style option Expected SameValue(«1 yr, 2 mths, 3 wks, 3 days, 4:05:06», «1 yr, 2 mths, 3 wks, 3 days, 4:05:06.007008009») to be true'
```
I'm slightly confused; this test was passing but it was passing for the wrong reason?

sosukesuzuki replied:
Yes, because this patch changed the behavior of the partitionDurationFormatPattern harness (https://github.com/WebKit/WebKit/pull/29128/files#r1616903101). This changed the expected value of this test. But the actual value did not change.
I will fix the actual value in a future PR.

rkirsling commented on Source/JavaScriptCore/runtime/IntlDurationFormat.cpp:
```diff
@@ -239,7 +239,8 @@ void IntlDurationFormat::initializeDurationFormat(JSGlobalObject* globalObject,
     }
 
-    m_fractionalDigits = intlNumberOption(globalObject, options, vm.propertyNames->fractionalDigits, 0, 9, 0);
+    constexpr unsigned undefinedValue = UINT32_MAX;
```
We should use `std::numeric_limits<unsigned>::max()` instead of `UINT32_MAX`. Also, you can just define this once at the top of the file, like [this](https://github.com/WebKit/WebKit/blob/main/Source/JavaScriptCore/runtime/TemporalDuration.cpp#L38).

sosukesuzuki replied:
Thank you for the review, I've fixed!

sosukesuzuki committed:
[JSC] `Intl.DurationFormat#resolvedOptions().fractionalDigits` should be `undefined` if `fractionalDigits` is `undefined`.
https://bugs.webkit.org/show_bug.cgi?id=274733

Reviewed by Ross Kirsling.

According to the spec[1], when creating a `DurationFormat` instance with
`new Intl.DurationFormat(locales, { fractionalDigits: undefined });`,
`Intl.DurationFormat#resolvedOptions().fractionalDigits` should be `undefined`. However, in the
current JSC, it is `0`. This was occurring because, according to the spec[2], the
`fallback` argument of the `GetNumberOption` abstract operation should be `undefined`, but we were
passing `0`.

This patch makes the following changes:

- Pass `numeric_limits<unsigned>::max()`, representing `undefined`, to the `fallback` argument of
the `intlNumberOption`
function, which corresponds to the `GetNumberOption` abstract operation, when initializing
`DurationFormat` instances.
- In the `DurationFormat#resolvedOption` function, set the `fractionalDigits` field to
`jsUndefined()` if `m_fractionalDigits` is `numeric_limits<unsigned>::max()`.

[1]: https://tc39.es/proposal-intl-duration-format/#sec-Intl.DurationFormat.prototype.resolvedOptions
[2]: https://tc39.es/proposal-intl-duration-format/#sec-Intl.DurationFormat

* JSTests/stress/intl-durationformat-fractionaldigits-undefined.js: Added.
(shouldBe):
* JSTests/stress/intl-durationformat.js:
(test):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/IntlDurationFormat.cpp:
(JSC::IntlDurationFormat::initializeDurationFormat):
(JSC::collectElements):
(JSC::IntlDurationFormat::resolvedOptions const):

Canonical link: https://commits.webkit.org/279503@main
